Sources/SourceKitLSP/Workspace.swift

```diff
@@ -108,20 +108,40 @@ public final class Workspace {
     reloadPackageStatusCallback: @escaping (ReloadPackageStatus) async -> Void
   ) async throws {
     var buildSystem: BuildSystem? = nil
-    if let rootUrl = rootUri.fileURL, let rootPath = try? AbsolutePath(validating: rootUrl.path) {
-      if let buildServer = await BuildServerBuildSystem(projectRoot: rootPath, buildSetup: buildSetup) {
-        buildSystem = buildServer
-      } else if let swiftpm = await SwiftPMWorkspace(
+
+    func buildSwiftPMWorkspace(rootUrl: URL) async -> SwiftPMWorkspace? {
+      return await SwiftPMWorkspace(
         url: rootUrl,
         toolchainRegistry: toolchainRegistry,
         buildSetup: buildSetup,
         reloadPackageStatusCallback: reloadPackageStatusCallback
-      ) {
-        buildSystem = swiftpm
-      } else if let compdb = CompilationDatabaseBuildSystem(
+      )
+    }
+
+    func buildCompDBWorkspace(rootPath: AbsolutePath) -> CompilationDatabaseBuildSystem? {
+      return CompilationDatabaseBuildSystem(
         projectRoot: rootPath,
         searchPaths: compilationDatabaseSearchPaths
-      ) {
+      )
+    }
+
+    func buildBuildServerWorkspace(rootPath: AbsolutePath) async -> BuildServerBuildSystem? {
+      return await BuildServerBuildSystem(projectRoot: rootPath, buildSetup: buildSetup)
+    }
+
+    if let rootUrl = rootUri.fileURL, let rootPath = try? AbsolutePath(validating: rootUrl.path) {
+      if let defaultWorkspaceType = buildSetup.defaultWorkspaceType, let defaultBuildSystem: BuildSystem? =
+        switch defaultWorkspaceType {
+          case .buildserver: await buildBuildServerWorkspace(rootPath: rootPath)
+          case .compdb: buildCompDBWorkspace(rootPath: rootPath)
+          case .swiftpm: await buildSwiftPMWorkspace(rootUrl: rootUrl)
+      } {
```

Comment:
I’m not a fan of nesting a `switch` expression inside an `if` condition. I think it hurts readability. I would prefer something like

```suggestion
      let defaultBuildSystem: BuildSystem? =
        switch defaultWorkspaceType {
          case .buildserver: await buildBuildServerWorkspace(rootPath: rootPath)
          case .compdb: buildCompDBWorkspace(rootPath: rootPath)
          case .swiftpm: await buildSwiftPMWorkspace(rootUrl: rootUrl)
          case nil: nil
      }
      if let defaultBuildSystem  {
```

Reply:
Much better!